### PR TITLE
Add stateless route attribute documentation

### LINKED ISSUE #13254: [FrameworkBundle][HttpKernel] Add session usage reporting in stateless …
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#35732
| PR author(s) | @mtarld
| Merged in    | 5.1-dev

### LINKED ISSUE #13281: [Routing] Add stateless route attribute
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#35782
| PR author(s) | @mtarld
| Merged in    | 5.1-dev

### LINKED ISSUE #13464: [HttpFoundation][HttpKernel][Security] Improve UnexpectedSessionUsageEx…
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#36129
| PR author(s) | @mtarld
| Merged in    | 5.1-dev

### PR DESCRIPTION
**Pull request by mtarld**
Fixes #13281, Fixes #13254, Fixes #13464

### CODE REVIEW

File: routing.rst [hunk 1]
```diff
@@ -1805,6 +1805,82 @@ with a locale. This can be done by defining a different prefix for each locale
             ;
         };
 
+.. _stateless-routing:
+
+Stateless Routes
+----------------
+
+Routes can configure a `stateless` boolean option in order to make sure that the controller
```
Comment (OskarStark) [inline routing.rst]:
```suggestion
Routes can configure a ``stateless`` boolean option in order to make sure that the controller
```

**Comment by nicolas-grekas** [comment]
Should be updated to fix https://github.com/symfony/symfony-docs/issues/13464 also.

File: routing.rst [hunk 2]
```diff
@@ -1805,6 +1805,82 @@ with a locale. This can be done by defining a different prefix for each locale
             ;
         };
 
+.. _stateless-routing:
+
+Stateless Routes
+----------------
+
+Routes can configure a ``stateless`` boolean option in order to make sure that the controller
```
[the remark this answers is not in the document]
Comment (mtarld) [reply routing.rst:1813]: Thanks for the review @javiereguiluz 

> So this is more like "probably I won't use the session ... but maybe I use it, and that's fine

You are perfectly right, I shouldn't have written "in order to make sure" but maybe "in order to be notified" instead.

About the non-strictness of that attribute, it was actually a suggestion by @nicolas-grekas. 
In fact, using stateless routes shouldn't break applications in production mode. It's the developer's responsibility to ensure (thanks to the exceptions) that the session isn't used during the development process.
Therefore, you can't indeed use it "blindly" atm.

Is it clearer?

[the remark this answers is not in the document]
Comment (javiereguiluz) [reply routing.rst:1813]: Thanks for your comment. The explanation is clear now ... but I don't know when/how to use this option.

I thought it was ... *"I don't use the session"* ... but in fact it's like ... "*Maybe I'll use the session ... and maybe I won't*" 🤔   There's something I'm missing here.

[the remark this answers is not in the document]
Comment (mtarld) [reply routing.rst:1813]: IMO, it's more like, "_session shouldn't be used, and I'll be able to check it during my development process and be notified in logs in production (if I missed something)_"

File: routing.rst [hunk 3]
```diff
@@ -1805,6 +1805,82 @@ with a locale. This can be done by defining a different prefix for each locale
...
+
+.. tip::
+
+    When the application uses full "language + territory" locales (e.g. ``fr_FR``,
+    ``fr_BE``), if the URLs are the same in all related locales, routes can use
+    only the language part (e.g. ``fr``) to avoid repeating the same URLs.
+
+A common requirement for internationalized applications is to prefix all routes
+with a locale. This can be done by defining a different prefix for each locale
+(and setting an empty prefix for your default locale if you prefer it):
```
Comment (wouterj) [inline routing.rst:1881]: I'm not really sure how this is related to stateless routes? It seems better suited in e.g. the http_cache articles maybe?

File: routing.rst [hunk 4]
```diff
@@ -1805,6 +1805,82 @@ with a locale. This can be done by defining a different prefix for each locale
...
+        return function (RoutingConfigurator $routes) {
+            $routes->add('homepage', '/')
+                ->controller([MainController::class, 'homepage'])
+                ->stateless()
+            ;
+        };
+
+If a stateless declared route is using the session, the application will:
+- Throw an `UnexpectedSessionUsageException` when debugging is enabled
+- Log a warning when debugging is disabled.
```
Comment (wouterj) [inline routing.rst:1871]:
```suggestion
* Throw an :class:`Symfony\\Component\\HttpKernel\\Exception\\UnexpectedSessionUsageException` when debugging is enabled;
* Log a warning when debugging is disabled.
```

(just some minor formatting changes)

File: routing.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (wouterj) [reply routing.rst:1813]: I think this discussion has been resolved in https://github.com/symfony/symfony/pull/36321

**Comment by wouterj** [comment]
> Should be updated to fix #13464 also.

Fyi, I've updated the PR description to include this. I think the current text covers the behavior in this PR more enough, so no need to update the changes in this PR :)

File: routing.rst (hunk 3 above)
Comment (mtarld) [reply routing.rst]: Actually it was my mistake, this text shouldn't be there

**Review by wouterj — APPROVED** [review]

File: routing.rst [hunk 5]
```diff
@@ -1805,6 +1805,77 @@ with a locale. This can be done by defining a different prefix for each locale
             ;
         };
 
+.. _stateless-routing:
+
+Stateless Routes
+----------------
+
```
Comment (HeahDude) [inline routing.rst:1812]: We're also missing  a reason to use this feature. What about adding a short sentence as introduction?
Something like:
```rst
Sometimes, when an HTTP response should be cached, it is important to ensure
that can happen.
```
?

File: routing.rst [hunk 6]
```diff
@@ -1805,6 +1805,77 @@ with a locale. This can be done by defining a different prefix for each locale
...
+        return function (RoutingConfigurator $routes) {
+            $routes->add('homepage', '/')
+                ->controller([MainController::class, 'homepage'])
+                ->stateless()
+            ;
+        };
+
+If a stateless declared route is using the session, the application will:
+- Throw an `Symfony\\Component\\HttpKernel\\Exception\\UnexpectedSessionUsageException` when debugging is enabled
+- Log a warning when debugging is disabled.
```
Comment (HeahDude) [inline routing.rst:1871]:
```suggestion
Now, if the session is used, the application will report it based on your
``kernel.debug`` parameter:
- ``enabled``: will throw an :class:`Symfony\\Component\\HttpKernel\\Exception\\UnexpectedSessionUsageException` exception
- ``disabled``: will log a warning
```

File: routing.rst [hunk 7]
```diff
@@ -1805,6 +1805,77 @@ with a locale. This can be done by defining a different prefix for each locale
...
+
+.. tip::
```
Comment (HeahDude) [inline routing.rst:1873]: This tip seems to not belong to this PR? Should target 4.4?

**Review by HeahDude — CHANGES_REQUESTED** [review]
Thank you very much for this great feature and documenting it! I propose some tweaks though, let me know what you think.

File: routing.rst (hunk 7 above)
Comment (mtarld) [reply routing.rst:1874]: My mistake :)

File: routing.rst (hunk 5 above)
Comment (mtarld) [reply routing.rst:1812]: I'd rather add more details for the reason why session is involved: 
```
Sometimes, when an HTTP response should be cached, it is important to ensure
that can happen. However, whenever session is started during a request, Symfony
turns the response into a private non-cacheable response.
```

WDYT ?

**Review by wouterj — APPROVED** [review]

File: routing.rst (hunk 5 above)
Comment (HeahDude) [reply routing.rst:1812]: Sounds good to me!

File: routing.rst [hunk 8]
```diff
@@ -1805,6 +1805,82 @@ with a locale. This can be done by defining a different prefix for each locale
...
+
+.. versionadded:: 5.1
+
+    The ``stateless`` option was introduced in Symfony 5.1
+
+Sometimes, when an HTTP response should be cached, it is important to ensure
+that can happen. However, whenever session is started during a request, Symfony
+turns the response into a private non-cacheable response.
+
+For details, see :doc:`/http_cache`
```
Comment (HeahDude) [inline routing.rst:1821]: Missing a dot.

File: routing.rst [hunk 9]
```diff
@@ -1805,6 +1805,82 @@ with a locale. This can be done by defining a different prefix for each locale
...
+
+Routes can configure a ``stateless`` boolean option in order to make sure that
+the session won't ever be used when matching a request:
```
Comment (HeahDude) [inline routing.rst:1824]:
```suggestion
Routes can configure a ``stateless`` boolean option in order to declare that the
session shouldn't be used when matching a request:
```

File: routing.rst [hunk 10]
```diff
@@ -1805,6 +1805,82 @@ with a locale. This can be done by defining a different prefix for each locale
...
+                ->controller([MainController::class, 'homepage'])
+                ->stateless()
+            ;
+        };
+
+Now, if the session is used, the application will report it based on your
+``kernel.debug`` parameter:
+* ``enabled``: will throw an :class:`Symfony\\Component\\HttpKernel\\Exception\\UnexpectedSessionUsageException` exception
+* ``disabled``: will log a warning
+
```
Comment (HeahDude) [inline routing.rst:1883]:
```rst
It well help you understanding and hopefully fixing unexpected behavior in your application.
```
To conclude?

**Review by HeahDude — COMMENTED** [review]
Looks great, thanks! Again some minor comments, let me know

File: routing.rst (hunk 10 above)
Comment (mtarld) [reply routing.rst:1883]: Good idea yes!

File: routing.rst [hunk 11]
```diff
@@ -1805,6 +1805,84 @@ with a locale. This can be done by defining a different prefix for each locale
...
         };
 
+.. _stateless-routing:
+
+Stateless Routes
+----------------
+
+.. versionadded:: 5.1
+
+    The ``stateless`` option was introduced in Symfony 5.1
```
Comment (HeahDude) [inline routing.rst:1815]: Missing a dot here too (last comment from me :)

**Review by HeahDude — APPROVED** [review]
Thanks!

**Comment by HeahDude** [comment]
Thank you Mathias.
